File: src/resources/server_info/SdbLoader.java
/************************************************************************************
 * Copyright (c) 2015 /// Project SWG /// www.projectswg.com                        *
 *                                                                                  *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on           *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies.  *
 * Our goal is to create an emulator which will provide a server for players to     *
 * continue playing a game similar to the one they used to play. We are basing      *
 * it on the final publish of the game prior to end-game events.                    *
 *                                                                                  *
 * This file is part of Holocore.                                                   *
 *                                                                                  *
 * -------------------------------------------------------------------------------- *
 *                                                                                  *
 * Holocore is free software: you can redistribute it and/or modify                 *
 * it under the terms of the GNU Affero General Public License as                   *
 * published by the Free Software Foundation, either version 3 of the               *
 * License, or (at your option) any later version.                                  *
 *                                                                                  *
 * Holocore is distributed in the hope that it will be useful,                      *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                   *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                    *
 * GNU Affero General Public License for more details.                              *
 *                                                                                  *
 * You should have received a copy of the GNU Affero General Public License         *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.                *
 *                                                                                  *
 ***********************************************************************************/
package resources.server_info;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

import com.projectswg.common.debug.Log;

public class SdbLoader {
	
	private SdbLoader() {
		
	}
	
	private enum DataType {
		TEXT,
		INTEGER,
		REAL,
		BOOLEAN;
		
		public Object decode(String str) {
			switch (this) {
				case TEXT:
				default:
					return str;
				case INTEGER:
					return Long.parseLong(str);
				case REAL:
					return Double.parseDouble(str);
				case BOOLEAN:
					return str.equalsIgnoreCase("true") || str.equals("1");
			}
		}
	}
	
	public static SdbResultSet load(File file) throws IOException {
		String ext = getExtension(file);
		if (ext.equals("msdb")) {
			return MasterSdbResultSet.load(file);
		} else if (ext.equals("sdb")) {
			return SingleSdbResultSet.load(file);
		} else {
			throw new IllegalArgumentException("Invalid file! Expected either msdb or sdb");
		}
	}
	
	private static String getExtension(File file) {
		String ext = file.getName().toLowerCase(Locale.US);
		int lastPeriod = ext.lastIndexOf('.');
		if (lastPeriod == -1)
			return ext;
		return ext.substring(lastPeriod+1);
	}
	
	public interface SdbResultSet extends AutoCloseable {
		
		@Override
		void close() throws IOException;
		boolean next() throws IOException;
		List<String> getColumns();
		
		Object getObject(int index);
		
		String getText(int index);
		String getText(String columnName);
		
		long getInt(int index);
		long getInt(String columnName);
		
		double getReal(int index);
		double getReal(String columnName);
		
		boolean getBoolean(int index);
		boolean getBoolean(String columnName);
	}
	
	private static class MasterSdbResultSet implements SdbResultSet {
		
		private final Iterator<SdbResultSet> sdbs;
		private final AtomicReference<SdbResultSet> sdb;
		
		private MasterSdbResultSet(Iterator<SdbResultSet> sdbs) {
			this.sdbs = sdbs;
			this.sdb = new AtomicReference<>(sdbs.hasNext() ? sdbs.next() : null);
		}
		
		@Override
		public void close() throws IOException {
			SdbResultSet set = getResultSet();
			if (set != null)
				set.close();
		}
		
		@Override
		public boolean next() throws IOException {
			SdbResultSet set = getResultSet();
			if (set == null)
				return false;
			while (!set.next()) {
				set.close();
				if (!sdbs.hasNext()) {
					sdb.set(null);
					return false; // bummer
				}
				set = sdbs.next();
				sdb.set(set);
				if (set == null)
					return false; // shouldn't be possible anyways
			}
			return true;
		}
		
		@Override
		public List<String> getColumns() {
			return getResultSet().getColumns();
		}
		
		@Override
		public Object getObject(int index) {
			return getResultSet().getObject(index);
		}
		
		@Override
		public String getText(int index) {
			return getResultSet().getText(index);
		}
		
		@Override
		public String getText(String columnName) {
			return getResultSet().getText(columnName);
		}
		
		@Override
		public long getInt(int index) {
			return getResultSet().getInt(index);
		}
		
		@Override
		public long getInt(String columnName) {
			return getResultSet().getInt(columnName);
		}
		
		@Override
		public double getReal(int index) {
			return getResultSet().getReal(index);
		}
		
		@Override
		public double getReal(String columnName) {
			return getResultSet().getReal(columnName);
		}
		
		@Override
		public boolean getBoolean(int index) {
			return getResultSet().getBoolean(index);
		}
		
		@Override
		public boolean getBoolean(String columnName) {
			return getResultSet().getBoolean(columnName);
		}
		
		private SdbResultSet getResultSet() {
			return sdb.get();
		}
		
		private static MasterSdbResultSet load(File file) throws IOException {
			List<SdbResultSet> sets = new ArrayList<>();
			File parentFile = file.getParentFile();
			try (SingleSdbResultSet msdb = SingleSdbResultSet.load(file)) {
				while (msdb.next()) {
					if (msdb.getBoolean(1)) // is enabled
						sets.add(SdbLoader.load(new File(parentFile, msdb.getText(0)))); // relative file path
				}
			}
			return new MasterSdbResultSet(sets.iterator());
		}
		
	}
	
	private static class SingleSdbResultSet implements SdbResultSet {
		
		private final File file;
		private final Map<String, Integer> columnNames;
		private final AtomicLong lineNumber;
		private DataType [] columnTypes;
		private Object [] columnValues;
		private BufferedReader reader;
		
		private SingleSdbResultSet(File file) {
			this.file = file;
			this.columnNames = new HashMap<>();
			this.lineNumber = new AtomicLong(0);
			this.columnTypes = null;
			this.columnValues = null;
			this.reader = null;
		}
		
		@Override
		public void close() throws IOException {
			if (reader != null)
				reader.close();
			reader = null;
		}
		
		@Override
		public boolean next() throws IOException {
			String line;
			do {
				line = reader.readLine();
			} while (line != null && line.isEmpty());
			
			if (line != null) {
				readNextLine(line);
				return true;
			}
			return false;
		}
		
		@Override
		public List<String> getColumns() {
			return new ArrayList<>(columnNames.keySet());
		}
		
		@Override
		public Object getObject(int index) {
			return columnValues[index];
		}
		
		@Override
		public String getText(int index) {
			return (String) columnValues[index];
		}
		
		@Override
		public String getText(String columnName) {
			return getText(columnNames.get(columnName));
		}
		
		@Override
		public long getInt(int index) {
			return (Long) columnValues[index];
		}
		
		@Override
		public long getInt(String columnName) {
			return getInt(columnNames.get(columnName));
		}
		
		@Override
		public double getReal(int index) {
			return (Double) columnValues[index];
		}
		
		@Override
		public double getReal(String columnName) {
			return getReal(columnNames.get(columnName));
		}
		
		@Override
		public boolean getBoolean(int index) {
			return (Boolean) columnValues[index];
		}
		
		@Override
		public boolean getBoolean(String columnName) {
			return getBoolean(columnNames.get(columnName));
		}
		
		private void readNextLine(String line) {
			long lineNum = lineNumber.incrementAndGet();
			int prevIndex = 0;
			int nextIndex = 0;
			int i = 0;
			try {
				for (i = 0; i < columnTypes.length-1; i++) {
					nextIndex = line.indexOf('\t', prevIndex);
					columnValues[i] = columnTypes[i].decode(line.substring(prevIndex, nextIndex));
					prevIndex = nextIndex+1;
				}
				columnValues[i] = columnTypes[i].decode(line.substring(prevIndex));
			} catch (NumberFormatException e) {
				throw new NumberFormatException("Failed to parse value in sdb: " + file + " on line " + lineNum + " in column " + (i+1));
			}
		}
		
		private static SingleSdbResultSet load(File file) throws IOException {
			SingleSdbResultSet sdb = new SingleSdbResultSet(file);
			sdb.load();
			return sdb;
		}
		
		private void load() throws IOException {
			reader = new BufferedReader(new FileReader(file));
			if (!loadHeader(reader.readLine(), reader.readLine()))
				return;
		}
		
		private boolean loadHeader(String columnsStr, String typesStr) {
			if (columnsStr == null || typesStr == null) {
				Log.e("Invalid SDB header: %s - nonexistent", file);
				return false;
			}
			String [] columns = columnsStr.split("\t");
			String [] types = typesStr.split("\t");
			if (columns.length != types.length) {
				Log.e("Invalid SDB header: %s - invalid lengths!", file);
				return false;
			}
			columnTypes = new DataType[columns.length];
			columnValues = new Object[columns.length];
			for (int i = 0; i < columns.length; i++) {
				columnTypes[i] = parseColumnType(types[i]);
				columnNames.put(columns[i], i);
			}
			lineNumber.set(2);
			return true;
		}
		
		private DataType parseColumnType(String type) {
			if (type.indexOf(' ') != -1)
				type = type.substring(0, type.indexOf(' '));
			if (type.equalsIgnoreCase("TEXT"))
				return DataType.TEXT;
			if (type.equalsIgnoreCase("INTEGER"))
				return DataType.INTEGER;
			if (type.equalsIgnoreCase("REAL"))
				return DataType.REAL;
			if (type.equalsIgnoreCase("BOOL") || type.equalsIgnoreCase("BOOLEAN"))
				return DataType.BOOLEAN;
			Log.e("Unknown column type: %s for file %s", type, file);
			return DataType.TEXT;
		}
		
	}
	
}

File: src/main/java/com/projectswg/holocore/resources/support/data/server_info/SdbLoader.kt
/***********************************************************************************
 * Copyright (c) 2024 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.support.data.server_info

import com.projectswg.holocore.resources.support.data.server_info.SdbColumnArraySet.*
import me.joshlarson.jlcommon.log.Log
import org.intellij.lang.annotations.Language
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.Spliterators.AbstractSpliterator
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Consumer
import java.util.function.Function
import java.util.stream.Stream
import java.util.stream.StreamSupport

object SdbLoader {
	@Throws(IOException::class)
	fun load(file: File): SdbResultSet {
		val ext = getExtension(file)
		return when (ext) {
			"msdb" -> MasterSdbResultSet.load(file)
			"sdb"  -> SingleSdbResultSet.load(file)
			else   -> throw IllegalArgumentException("Invalid file! Expected either msdb or sdb")
		}
	}

	private fun getExtension(file: File): String {
		val ext = file.name.lowercase()
		val lastPeriod = ext.lastIndexOf('.')
		if (lastPeriod == -1) return ext
		return ext.substring(lastPeriod + 1)
	}

	interface SdbResultSet : Closeable, AutoCloseable {
		@Throws(IOException::class)
		override fun close()
		@Throws(IOException::class)
		fun next(): Boolean
		val columns: List<String>
		val file: File
		val line: Int

		/**
		 * Returns a sequential stream after applying the specified transformation to the SdbResultSet
		 * @param transform the transformation to apply
		 * @param <T> the type of the returned stream
		 * @return a new sequential stream to iterate over the SDB
		</T> */
		fun <T> stream(transform: Function<SdbResultSet, T>): Stream<T>

		/**
		 * Returns a parallel stream after applying the specified transformation to the SdbResultSet
		 * @param transform the transformation to apply
		 * @param <T> the type of the returned stream
		 * @return a new parallel stream to iterate over the SDB
		</T> */
		fun <T> parallelStream(transform: Function<SdbResultSet, T>): Stream<T>

		fun getTextArrayParser(@Language("RegExp") regex: String, defValue: String?): SdbTextColumnArraySet
		fun getIntegerArrayParser(@Language("RegExp") regex: String, defValue: Int): SdbIntegerColumnArraySet
		fun getLongArrayParser(@Language("RegExp") regex: String, defValue: Long): SdbLongColumnArraySet
		fun getRealArrayParser(@Language("RegExp") regex: String, defValue: Double): SdbRealColumnArraySet
		fun getBooleanArrayParser(@Language("RegExp") regex: String, defValue: Boolean): SdbBooleanColumnArraySet

		fun getText(index: Int): String
		fun getText(columnName: String): String

		fun getInt(index: Int): Long
		fun getInt(columnName: String): Long

		fun getReal(index: Int): Double
		fun getReal(columnName: String): Double

		fun getBoolean(index: Int): Boolean
		fun getBoolean(columnName: String): Boolean
	}

	private class MasterSdbResultSet private constructor(private val sdbList: List<SdbResultSet>) : SdbResultSet {
		private val sdbs: Iterator<SdbResultSet> = sdbList.iterator()
		private val sdb: AtomicReference<SdbResultSet?> = AtomicReference(if (sdbs.hasNext()) sdbs.next() else null)

		@Throws(IOException::class)
		override fun close() {
			resultSet?.close()
		}

		override fun <T> stream(transform: Function<SdbResultSet, T>): Stream<T> {
			return stream(transform, false)
		}

		override fun <T> parallelStream(transform: Function<SdbResultSet, T>): Stream<T> {
			return stream(transform, true)
		}

		fun <T> stream(transform: Function<SdbResultSet, T>, parallel: Boolean): Stream<T> {
			return StreamSupport.stream(Spliterators.spliterator(sdbList, Spliterator.NONNULL or Spliterator.IMMUTABLE or Spliterator.DISTINCT), parallel).flatMap { sdb: SdbResultSet -> sdb.stream(transform) }
		}

		@Throws(IOException::class)
		override fun next(): Boolean {
			var set = resultSet ?: return false
			while (!set.next()) {
				set.close()
				if (!sdbs.hasNext()) {
					sdb.set(null)
					return false // bummer
				}
				set = sdbs.next()
				sdb.set(set)
			}
			return true
		}

		override val columns: List<String>
			get() = resultSet!!.columns

		override val file: File
			get() = resultSet!!.file

		override val line: Int
			get() = resultSet!!.line

		override fun getTextArrayParser(@Language("RegExp") regex: String, defValue: String?): SdbTextColumnArraySet {
			return SdbTextColumnArraySet(regex, defValue)
		}

		override fun getIntegerArrayParser(@Language("RegExp") regex: String, defValue: Int): SdbIntegerColumnArraySet {
			return SdbIntegerColumnArraySet(regex, defValue)
		}

		override fun getLongArrayParser(@Language("RegExp") regex: String, defValue: Long): SdbLongColumnArraySet {
			return SdbLongColumnArraySet(regex, defValue)
		}

		override fun getRealArrayParser(@Language("RegExp") regex: String, defValue: Double): SdbRealColumnArraySet {
			return SdbRealColumnArraySet(regex, defValue)
		}

		override fun getBooleanArrayParser(@Language("RegExp") regex: String, defValue: Boolean): SdbBooleanColumnArraySet {
			return SdbBooleanColumnArraySet(regex, defValue)
		}

		override fun getText(index: Int): String {
			return resultSet!!.getText(index)
		}

		override fun getText(columnName: String): String {
			return resultSet!!.getText(columnName)
		}

		override fun getInt(index: Int): Long {
			return resultSet!!.getInt(index)
		}

		override fun getInt(columnName: String): Long {
			return resultSet!!.getInt(columnName)
		}

		override fun getReal(index: Int): Double {
			return resultSet!!.getReal(index)
		}

		override fun getReal(columnName: String): Double {
			return resultSet!!.getReal(columnName)
		}

		override fun getBoolean(index: Int): Boolean {
			return resultSet!!.getBoolean(index)
		}

		override fun getBoolean(columnName: String): Boolean {
			return resultSet!!.getBoolean(columnName)
		}

		private val resultSet: SdbResultSet?
			get() = sdb.get()

		companion object {
			@Throws(IOException::class)
			fun load(file: File): MasterSdbResultSet {
				val sets: MutableList<SdbResultSet> = ArrayList()
				val parentFile = file.parentFile
				SingleSdbResultSet.load(file).use { msdb ->
					while (msdb.next()) {
						if (msdb.getBoolean(1)) // is enabled
							sets.add(SdbLoader.load(File(parentFile, msdb.getText(0)))) // relative file path
					}
				}
				return MasterSdbResultSet(sets)
			}
		}
	}

	private class SingleSdbResultSet private constructor(override val file: File) : SdbResultSet {
		private val columnIndices: MutableMap<String, Int> = HashMap()
		private val lineNumber = AtomicLong(0)
		private var columnNames: Array<String>?
		private var columnValues: Array<String>?
		private var input: BufferedReader?

		init {
			this.columnNames = null
			this.columnValues = null
			this.input = null
		}

		@Throws(IOException::class)
		override fun close() {
			input!!.close()
		}

		override fun <T> stream(transform: Function<SdbResultSet, T>): Stream<T> {
			return stream(transform, false)
		}

		override fun <T> parallelStream(transform: Function<SdbResultSet, T>): Stream<T> {
			return stream(transform, true)
		}

		fun <T> stream(transform: Function<SdbResultSet, T>, parallel: Boolean): Stream<T> {
			val resultSet = ThreadLocal.withInitial { ParallelSdbResultSet(file, columnIndices, columnNames) }
			return StreamSupport.stream(SdbSpliterator(file, input, lineNumber), parallel).map { e: Map.Entry<Long, String> ->
				val set = resultSet.get()
				set.load(e.value, e.key)
				transform.apply(set)
			}
		}

		override fun next(): Boolean {
			var line: String?
			do {
				lineNumber.incrementAndGet()
				try {
					line = input!!.readLine()
					if (line == null) return false
				} catch (e: IOException) {
					return false
				}
			} while (line!!.isEmpty())
			var index = 0
			val columnCount = columnValues!!.size
			for (column in 0 until columnCount) {
				var nextIndex = line.indexOf('\t', index)
				if (nextIndex == -1) {
					if (column + 1 < columnCount) {
						Log.e("Invalid entry in sdb: %s on line %d - invalid number of columns!", file, lineNumber.get(), column + 1)
						return false
					}
					nextIndex = line.length
				}
				columnValues!![column] = line.substring(index, nextIndex)
				index = nextIndex + 1
			}
			return true
		}

		override val columns: List<String>
			get() = columnNames?.toList() ?: listOf()

		override val line: Int
			get() = lineNumber.get().toInt()

		override fun getTextArrayParser(@Language("RegExp") regex: String, defValue: String?): SdbTextColumnArraySet {
			return SdbTextColumnArraySet(regex, defValue)
		}

		override fun getIntegerArrayParser(@Language("RegExp") regex: String, defValue: Int): SdbIntegerColumnArraySet {
			return SdbIntegerColumnArraySet(regex, defValue)
		}

		override fun getLongArrayParser(@Language("RegExp") regex: String, defValue: Long): SdbLongColumnArraySet {
			return SdbLongColumnArraySet(regex, defValue)
		}

		override fun getRealArrayParser(@Language("RegExp") regex: String, defValue: Double): SdbRealColumnArraySet {
			return SdbRealColumnArraySet(regex, defValue)
		}

		override fun getBooleanArrayParser(@Language("RegExp") regex: String, defValue: Boolean): SdbBooleanColumnArraySet {
			return SdbBooleanColumnArraySet(regex, defValue)
		}

		override fun getText(index: Int): String {
			return columnValues!![index]
		}

		override fun getText(columnName: String): String {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist in sdb $file" }
			return columnValues!![columnIndices[columnName]!!]
		}

		override fun getInt(index: Int): Long {
			try {
				return columnValues!![index].toLong()
			} catch (e: NumberFormatException) {
				throw NumberFormatException("Failed to parse value in sdb: " + file + " on line " + lineNumber.get() + " in column " + (index + 1))
			}
		}

		override fun getInt(columnName: String): Long {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist in sdb $file" }
			return getInt(columnIndices[columnName]!!)
		}

		override fun getReal(index: Int): Double {
			try {
				return columnValues!![index].toDouble()
			} catch (e: NumberFormatException) {
				throw NumberFormatException("Failed to parse value in sdb: " + file + " on line " + lineNumber.get() + " in column " + (index + 1))
			}
		}

		override fun getReal(columnName: String): Double {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist in sdb $file" }
			return getReal(columnIndices[columnName]!!)
		}

		override fun getBoolean(index: Int): Boolean {
			return columnValues!![index].equals("true", ignoreCase = true)
		}

		override fun getBoolean(columnName: String): Boolean {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist in sdb $file" }
			return getBoolean(columnIndices[columnName]!!)
		}

		@Throws(IOException::class)
		private fun load() {
			input = BufferedReader(FileReader(file, StandardCharsets.UTF_8), 128 * 1024)
			loadHeader(input!!.readLine())
			input!!.readLine()
		}

		private fun loadHeader(columnsStr: String?) {
			if (columnsStr == null) {
				Log.e("Invalid SDB header: %s - nonexistent", file)
				return
			}
			columnNames = columnsStr.split("\t".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
			columnValues = Array(columnNames!!.size) { "" }
			for (i in columnNames!!.indices) {
				columnIndices[columnNames!![i]] = i
			}
			lineNumber.set(2)
		}

		companion object {
			@Throws(IOException::class)
			fun load(file: File): SingleSdbResultSet {
				val sdb = SingleSdbResultSet(file)
				sdb.load()
				return sdb
			}
		}
	}

	private class ParallelSdbResultSet(override val file: File, private val columnIndices: Map<String, Int>, private val columnNames: Array<String>?) : SdbResultSet {
		private val lineNumber = AtomicLong(0)
		private val columnValues: Array<String> = Array(columnIndices.size) { "" }

		override fun close() {
			throw UnsupportedOperationException("Cannot close a parallel sdb")
		}

		override fun next(): Boolean {
			throw UnsupportedOperationException("Cannot iterate a parallel sdb")
		}

		override fun <T> stream(transform: Function<SdbResultSet, T>): Stream<T> {
			throw UnsupportedOperationException("Cannot iterate a parallel sdb")
		}

		override fun <T> parallelStream(transform: Function<SdbResultSet, T>): Stream<T> {
			throw UnsupportedOperationException("Cannot iterate a parallel sdb")
		}

		fun load(line: String, lineNumber: Long) {
			var index = 0
			val columnCount = columnValues.size
			for (column in 0 until columnCount) {
				var nextIndex = line.indexOf('\t', index)
				if (nextIndex == -1) {
					if (column + 1 < columnCount) {
						Log.e("Invalid entry in sdb: %s on line %d - invalid number of columns!", file, lineNumber, column + 1)
						return
					}
					nextIndex = line.length
				}
				columnValues[column] = line.substring(index, nextIndex)
				index = nextIndex + 1
			}
		}

		override val columns: List<String>
			get() = columnNames?.toList() ?: listOf()

		override val line: Int
			get() = lineNumber.get().toInt()

		override fun getTextArrayParser(@Language("RegExp") regex: String, defValue: String?): SdbTextColumnArraySet {
			return SdbTextColumnArraySet(regex, defValue)
		}

		override fun getIntegerArrayParser(@Language("RegExp") regex: String, defValue: Int): SdbIntegerColumnArraySet {
			return SdbIntegerColumnArraySet(regex, defValue)
		}

		override fun getLongArrayParser(@Language("RegExp") regex: String, defValue: Long): SdbLongColumnArraySet {
			return SdbLongColumnArraySet(regex, defValue)
		}

		override fun getRealArrayParser(@Language("RegExp") regex: String, defValue: Double): SdbRealColumnArraySet {
			return SdbRealColumnArraySet(regex, defValue)
		}

		override fun getBooleanArrayParser(@Language("RegExp") regex: String, defValue: Boolean): SdbBooleanColumnArraySet {
			return SdbBooleanColumnArraySet(regex, defValue)
		}

		override fun getText(index: Int): String {
			return columnValues[index]
		}

		override fun getText(columnName: String): String {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist" }
			return columnValues[columnIndices[columnName]!!]
		}

		override fun getInt(index: Int): Long {
			try {
				return columnValues[index].toLong()
			} catch (e: NumberFormatException) {
				throw NumberFormatException("Failed to parse value in sdb: " + file + " on line " + lineNumber.get() + " in column " + (index + 1))
			}
		}

		override fun getInt(columnName: String): Long {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist" }
			return getInt(columnIndices[columnName]!!)
		}

		override fun getReal(index: Int): Double {
			try {
				return columnValues[index].toDouble()
			} catch (e: NumberFormatException) {
				throw NumberFormatException("Failed to parse value in sdb: " + file + " on line " + lineNumber.get() + " in column " + (index + 1))
			}
		}

		override fun getReal(columnName: String): Double {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist" }
			return getReal(columnIndices[columnName]!!)
		}

		override fun getBoolean(index: Int): Boolean {
			return columnValues[index].equals("true", ignoreCase = true)
		}

		override fun getBoolean(columnName: String): Boolean {
			assert(columnIndices.containsKey(columnName)) { "column $columnName does not exist" }
			return getBoolean(columnIndices[columnName]!!)
		}
	}

	private class SdbSpliterator(private val file: File, private val input: BufferedReader?, private val lineNumber: AtomicLong) : AbstractSpliterator<Map.Entry<Long, String>>(Long.MAX_VALUE, ORDERED or NONNULL or IMMUTABLE or DISTINCT) {
		@Synchronized
		override fun tryAdvance(action: Consumer<in Map.Entry<Long, String>>): Boolean {
			var line: String
			var lineNumber: Long = -1
			try {
				synchronized(input!!) {
					lineNumber = this.lineNumber.incrementAndGet()
					line = input.readLine() ?: return false
				}
				if (line.isEmpty()) return true
				action.accept(java.util.Map.entry(lineNumber, line))
				return true
			} catch (e: IOException) {
				throw RuntimeException(e)
			} catch (t: Throwable) {
				if (t.cause == null) {
					Log.e("Failed to load line %d on SDB %s due to %s: %s", lineNumber, file, t.javaClass.name, t.message)
				} else {
					Log.e("Failed to load line %d on SDB %s", lineNumber, file)
					var temp: Throwable? = t
					while (temp != null) {
						Log.e("    %s: %s", temp.javaClass.name, temp.message)
						temp = temp.cause!!
					}
				}
				return true
			}
		}
	}
}
